File: CS315-Programming Languages/HW1/Codes/22101782_akin_yasemin.dart
// 22101782_akin_yasemin.dart

void main() {
  // 1. Boolean operators
  bool a = true;
  bool b = false;
  print("1. Boolean operators in Dart:");
  print("AND '&&' (a && b): ${a && b}"); // false
  print("OR '||' (a || b): ${a || b}"); // true
  print("NOT '!' (!a): ${!a}"); // false

  // 2. Data types for operands of boolean operators, Boolean values (Truthy and Falsy values)
  print("\n2. Data types for operands of boolean operators, Boolean values (Truthy and Falsy values) in Dart:");
  // Dart strictly uses bool type (has strict typing) for Boolean expressions, because Dart is a statically typed language and Non-Boolean values cannot be directly used as operasnd without explicit conversion.
  // The following lines will result in a compilation error in Dart
  // print("Number in boolean context:", 5);   // Compilation error
  // print("String in boolean context:", "hello"); // Compilation error
  // print("String in Boolean context: " + ("string" && a));
  // print("Number in Boolean context: " + (0 && a));

  // Correct way to use non-boolean values in boolean context
  int number = 5;
  String str = "hello";
  print("Is number non-zero?: ${number != 0}");
  print("Is string non-empty?: ${str.isNotEmpty}");

  // 3. Operator precedence rules
  print("\n3. Operator precedence rules in Dart: NOT > AND > OR");
  bool c = false;
  bool d = true;
  print("a || b && !c: ${a || b && !c}");                               // Equivalent to a || (b && (!c))
  print("a || b == c && d != a: ${a || b == c && d != a}");             // Equivalent to a || (b == c) && (d != a)
  print("!a && b || c == d && d != a: ${!a && b || c == d && d != a}"); // Equivalent to ((!a) && b) || ((c == d) && (d != a))

  // 4. Operator associativity rules
  print("\n4. Operator associativity rules in Dart: left-associative");
  print("a && b && c && d: ${a && b && c && d}"); // Evaluates as (((a && b) && c) && d)

  // 5. Operand evaluation order
  print("\n5. Operand evaluation order in Dart: left-to-right");
  bool result1 = firstMethod() && secondMethod();
  print(result1);
  bool result2 = secondMethod() || firstMethod();
  print(result2);

  // 6. Short-circuit evaluation
  print("\n6. Short-circuit evaluation in Dart:");
  print("false && secondMethod(): ${false && secondMethod()}"); // secondMethod is not called
  print("true || secondMethod(): ${true || secondMethod()}");   // secondMethod is not called
  bool result = b && (a || check());
  print("Result of b && (a || check()): $result"); //check() is not called because value of the second operand for the or operator does not matter at this example.
}

bool firstMethod() {
  print("First method called");
  return true;
}

bool secondMethod() {
  print("Second method called");
  return false;
}

bool check() {
  print("check() function called");
  return true;
}

File: CS315-Programming Languages/HW1/Codes/22101782_akin_yasemin.go
// 22101782_akin_yasemin.go

package main

import "fmt"

func main() {

	// 1. Boolean operators
	a := true
	b := false
	fmt.Println("1. Boolean operators in Go:")
	fmt.Println("AND '&&' (a && b):", a && b)        // false
	fmt.Println("OR '||' (a || b):", a || b)         // true
	fmt.Println("NOT '!' (!a):", !a)                 // false

	// 2. Data types for operands of boolean operators, Boolean values (Truthy and Falsy values)
	fmt.Println("\n2. Data types for operands of boolean operators, Boolean values (Truthy and Falsy values) in Go: no truthy or falsy values")
	fmt.Println("Boolean true:", true)   // true
	fmt.Println("Boolean false:", false) // false

	// Attempting to use non-boolean values in boolean context will result in a compilation error.
	var number int = 5
	var str string = "hello"
	var ptr *int = &number
	var slice []int = []int{1, 2, 3}

	// Uncommenting any of the following lines will result in a compilation error in Go
	// fmt.Println("Number in boolean context:", bool(number))
	// fmt.Println("String in boolean context:", bool(str))
	// fmt.Println("Pointer in boolean context:", bool(ptr))
	// fmt.Println("Slice in boolean context:", bool(slice))

	// Correct way to use non-boolean values in boolean context
	fmt.Println("Is number non-zero?:", number != 0)
	fmt.Println("Is string non-empty?:", str != "")
	fmt.Println("Is pointer non-nil?:", ptr != nil)
	fmt.Println("Is slice non-empty?:", len(slice) > 0)

	// 3. Operator precedence rules
	fmt.Println("\n3. Operator precedence rules in Go: NOT > AND > OR")
	c := false
	d := true
	fmt.Println("a || b && !c:", a || b && !c)                               // Equivalent to a || (b && (!c))
	fmt.Println("a || b == c && d != a:", a || b == c && d != a)             // Equivalent to a || (b == c) && (d != a)
	fmt.Println("!a && b || c == d && d != a:", !a && b || c == d && d != a) //Equivalent to ((!a) && b) || ((c == d) && (d != a))

	// 4. Operator associativity rules
	fmt.Println("\n4. Operator associativity rules in Go: left-associative")
	fmt.Println("a && b && c && d:", a && b && c && d) // Evaluates as (((a && b) && c) && d)
	x := 0
	y := 1
	fmt.Println("x > y || y > x || y == x:", x > y || y > x || y == x) // Evaluates as ((x > y || y > x) || y == x)
	
	// 5. Operand evaluation order
	fmt.Println("\n5. Operand evaluation order in Go: left-to-right")
	result1 := firstMethod() && secondMethod()
	fmt.Println(result1)
	result2 := secondMethod() || firstMethod()
	fmt.Println(result2)

	// 6. Short-circuit evaluation
	fmt.Println("\n6. Short-circuit evaluation in Go:")
	fmt.Println("false && secondMethod():", false && secondMethod()) // secondMethod is not called
	fmt.Println("true || secondMethod():", true || secondMethod())   // secondMethod is not called
}

func firstMethod() bool {
	fmt.Println("First method called")
	return true
}

func secondMethod() bool {
	fmt.Println("Second method called")
	return false
}